File: InternshipProjects/01_hospital_fixed_asset_tracker_project/source_code/products/AssetTrackerBackend/app/services/users/user_admin_policy.py
"""
File: user_admin_policy.py
Author: Farros Ramzy (you@domain.com)
Description: Safety rules for admin account management.
Version: 0.1
Date: 2026-06-05

Copyright (c) 2026
"""

from fastapi import HTTPException, status
from sqlalchemy import func

from app.db_models import User

ADMIN_ROLE = "admin"


def normalize_role(role: str | None) -> str:
    return str(role or "").strip().lower()


def is_admin(user: User | None) -> bool:
    return user is not None and normalize_role(user.role) == ADMIN_ROLE


def count_active_admins(db) -> int:
    return (
        db.query(User)
        .filter(func.lower(User.role) == ADMIN_ROLE)
        .filter(User.is_active.is_(True))
        .count()
    )


def is_last_active_admin(db, user: User) -> bool:
    return bool(is_admin(user) and user.is_active and count_active_admins(db) <= 1)


def ensure_can_change_role(db, target_user: User, current_user: User, new_role: str):
    if target_user.id == current_user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="cannot_change_own_role",
        )

    demoting_admin = is_admin(target_user) and normalize_role(new_role) != ADMIN_ROLE

    if demoting_admin and is_last_active_admin(db, target_user):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="cannot_demote_last_active_admin",
        )


def ensure_can_change_active(db, target_user: User, current_user: User, is_active: bool):
    if is_active:
        return

    if target_user.id == current_user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="admin_cannot_deactivate_self",
        )

    if is_last_active_admin(db, target_user):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="cannot_disable_last_active_admin",
        )


def ensure_can_admin_reset_password(target_user: User, current_user: User):
    if target_user.id == current_user.id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="cannot_reset_own_password_here",
        )

File: InternshipProjects/01_hospital_fixed_asset_tracker_project/source_code/products/AssetTrackerRegistrationFrontEnd/src/utils/userAdminPolicy.js
/**
 * @file userAdminPolicy.js
 * @author Farros Ramzy (you@domain.com)
 * @description Frontend safety rules for admin account actions.
 * @version 1.0.0
 * @date 2026-06-05

 * @Copyright (c) 2026
 */


export function normalizeRole(role) {
    return String(role || "").trim().toLowerCase().replace(/[\s-]+/g, "_");
}

export function isAdmin(user) {
    return normalizeRole(user?.role) === "admin";
}

export function getActiveAdminCount(users) {
    return users.filter((item) => isAdmin(item) && item.is_active).length;
}

export function getUserAdminPolicy({ targetUser, currentUser, users }) {
    const targetIsSelf = Number(targetUser?.id) === Number(currentUser?.id);
    const targetIsAdmin = isAdmin(targetUser);
    const activeAdminCount = getActiveAdminCount(users);
    const targetIsLastActiveAdmin = targetIsAdmin && targetUser?.is_active && activeAdminCount <= 1;

    const canDisable = Boolean(targetUser?.is_active) && !targetIsSelf && !targetIsLastActiveAdmin;
    const canEnable = Boolean(targetUser) && !targetUser?.is_active;
    const canToggleActive = canEnable || canDisable;
    const canResetPassword = Boolean(targetUser?.id) && !targetIsSelf;

    return {
        targetIsSelf,
        targetIsLastActiveAdmin,
        canToggleActive,
        canResetPassword,
        toggleReason: getToggleReason(targetIsSelf, targetIsLastActiveAdmin),
        resetReason: targetIsSelf ? "Use Change Password for your own account." : "",
    };
}

function getToggleReason(targetIsSelf, targetIsLastActiveAdmin) {
    if (targetIsSelf) {
        return "You cannot disable your own account.";
    }

    if (targetIsLastActiveAdmin) {
        return "At least one active admin is required.";
    }

    return "";
}
